feat: robots directive, full robots tag support

BrocksiNet opened the pull request:
### 1. Why is this change necessary?

The current robots.txt implementation in Shopware only supports basic `Allow` and `Disallow` directives without User-agent blocks, making it impossible to:
- Define bot-specific rules (e.g., different rules for Googlebot vs Bingbot)
- Use advanced directives like `Crawl-delay`, `Request-rate`, etc.
- Validate robots.txt syntax and catch configuration errors
- Extend robots.txt functionality through events

This limits SEO capabilities and prevents proper crawler management for e-commerce stores.

### 2. What does this change do, exactly?

**New Features:**
- ✅ **Full robots.txt standard support**: All 8 common directives (`User-agent`, `Allow`, `Disallow`, `Crawl-delay`, `Sitemap`, `Request-rate`, `Visit-time`, `Host`)
- ✅ **User-agent blocks**: Define bot-specific rules that are global across domains
- ✅ **Error detection & logging**: Parse validation with line-specific error/warning messages logged when configuration is saved
- ✅ **Extensibility events**: `RobotsDirectiveParsingEvent` and `RobotsUnknownDirectiveEvent` for custom directives and validation
- ✅ **Disable defaults option**: New config `core.basicInformation.robotsDisableDefaults` to disable Shopware's default rules

**Technical Changes:**
- Added `RobotsDirectiveParser` with full parsing, validation, and event dispatching
- Added value objects: `ParsedRobots`, `ParseIssue`, `ParseIssueSeverity`, `RobotsDirective`, `RobotsDirectiveType`, `RobotsUserAgentBlock`
- Added `RobotsConfigChangeSubscriber` for parse error logging
- Deprecated string format in `DomainRuleStruct` constructor (removed in v6.8.0)
- Enhanced `RobotsPageLoader` to separate global User-agent blocks from domain-specific path rules
- All parsing now requires `Context` parameter for proper event handling

**Backward Compatibility:**
- Existing configurations without User-agent directives work unchanged
- String format in `DomainRuleStruct` still works but triggers deprecation warning
- All new code paths tested with 105 unit tests

### 3. Describe each step to reproduce the issue or behaviour.

**Before (Issue):**
1. Go to Settings > Basic Information > Robots.txt
2. Try to add: `User-agent: Googlebot\nCrawl-delay: 10`
3. ❌ User-agent directive is ignored, crawl-delay has no effect

**After (Fixed):**
1. Go to Settings > Basic Information > Robots.txt
2. Add:
```
User-agent: Googlebot
Crawl-delay: 10
Disallow: /admin/
```
3. ✅ User-agent block appears in robots.txt with proper rules
4. ✅ Syntax errors are logged with line numbers when saved
5. ✅ Events allow custom validation and directive handling

### 4. Please link to the relevant issues (if any).

- Closes #12787 - Changes on the robots.txt dont allow "User-Agent" directives

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
  - 105 unit tests covering parsing, validation, events, and backward compatibility
- [x] I have created a [changelog file](https://github.com/shopware/shopware/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
  - `changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md`
- [x] I have written or adjusted the documentation according to my changes
  - Comprehensive upgrade information with examples in changelog
  - Migration guide for deprecated string format
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
  - All new classes have PHPDoc comments
  - Complex parsing logic documented
- [x] I have read the contribution requirements and fulfilled them
  - PSR-12 compliant
  - PHPStan clean
  - ECS clean
  - Backward compatible with deprecation path

BrocksiNet committed:
feat: robots directive, full robots tag support

BrocksiNet committed:
fix: remove breaks, ludtwig

BrocksiNet committed:
fix: bc breaks

BrocksiNet committed:
fix: add missing test

BrocksiNet committed:
fix: ecs for missing test

BrocksiNet committed:
fix: improve coverage RobotsPageLoader

BrocksiNet committed:
Merge branch 'trunk' into feat/improve-robots-txt

BrocksiNet committed:
fix: make setup for tests easier

BrocksiNet committed:
fix: ignore RobotsPage

mitelg commented on .bc-exclude.php:
```diff
@@ -128,5 +128,8 @@
         preg_quote('ADDED: Parameter domain was added to Method invalidDomain() of class Shopware\Core\System\SystemConfig\SystemConfigException', '/'),
+
+        // Type widening from string to ParsedRobots|string is backward compatible - all existing string usage continues to work
+        preg_quote('CHANGED: The parameter $rules of Shopware\Storefront\Page\Robots\Struct\DomainRuleStruct#__construct() changed from string to Shopware\Storefront\Page\Robots\Parser\ParsedRobots|string', '/'),
```
should not be necessary anymore since https://github.com/shopware/shopware/pull/12941

BrocksiNet replied:
sadly not working

mitelg commented on changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md:
```diff
@@ -0,0 +1,128 @@
+new DomainRuleStruct('Disallow: /admin/', '/en');
+
+// New signature (also works)
+new DomainRuleStruct($parsedRobots, '/en');
```
maybe add it like this and some example parameters?
```suggestion
$parsedRobots = new ParsedRobots();
new DomainRuleStruct($parsedRobots, '/en');
```

mitelg commented on changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md:
````diff
@@ -0,0 +1,128 @@
+new DomainRuleStruct($parsedRobots, '/en');
+```
+
+**This change is backward compatible** - all existing code passing a string continues to work without modification. The union type `ParsedRobots|string` accepts everything the old `string` type accepted, plus the new parsed object type for internal use.
````
I don't think we need to explain that again. you already state above, that you accept both types

mitelg commented on changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md:
```diff
@@ -0,0 +1,128 @@
+___
+# Upgrade Information
+
+## Technical Changes (Backward Compatible)
```
remember that this will be the headline in the upgrade file. 

`Technical Changes (Backward Compatible)` will be quite lost without its context. 

so probably, you should add a title for the whole topic and all `##` headlines should become `###` headlines

mitelg commented on src/Storefront/Page/Robots/Struct/DomainRuleStruct.php:
```diff
@@ -13,9 +14,22 @@ class DomainRuleStruct extends Struct
+    private array $directives = [];
+
+    /**
+     * @param ParsedRobots|string $rules The robots.txt rules as string (legacy) or parsed object (new)
```
do we want to deprecate the usage of a string? (if so, needs to be added to changelog as well)

mitelg commented on src/Core/System/Resources/config/basicInformation.xml:
```diff
@@ -204,12 +204,19 @@
+            <label>Disable default robots.txt rules</label>
+            <label lang="de-DE">Standard-Regeln für robots.txt deaktivieren</label>
+            <helpText>When enabled, Shopware's default robots.txt rules will not be rendered. Use this if you want full control over the robots.txt content.</helpText>
+            <helpText lang="de-DE">Wenn aktiviert, werden die Standard robots.txt-Regeln von Shopware nicht gerendert. Verwenden Sie dies, wenn Sie die volle Kontrolle über den robots.txt-Inhalt haben möchten.</helpText>
```
I had problems understanding what this options does and it only got clearer after I looked at the code. maybe it is possible to describe better what it does?

mitelg commented on src/Storefront/Page/Robots/Parser/ParsedRobots.php:
```diff
@@ -0,0 +1,42 @@
+     * @param RobotsDirective[] $orphanedPathDirectives
+     */
+    public function __construct(
+        private readonly array $userAgentBlocks,
```
why not public readonly properties?

mitelg commented on src/Storefront/Page/Robots/Parser/ParsedRobots.php:
```diff
@@ -0,0 +1,42 @@
+class ParsedRobots
+{
+    /**
+     * @param RobotsUserAgentBlock[] $userAgentBlocks
```
```suggestion
     * @param list<RobotsUserAgentBlock> $userAgentBlocks
```

list is more precise. `string[]` can also be `array<string, string>` or `array<int, string>`

please change every occurrence

mitelg commented on src/Storefront/Page/Robots/Parser/RobotsDirectiveParser.php:
```diff
@@ -0,0 +1,102 @@
+use Shopware\Storefront\Page\Robots\Struct\RobotsUserAgentBlock;
+
+#[Package('framework')]
+class RobotsDirectiveParser
```
sorry, I came up with that so late, but maybe we can use a library? :see_no_evil: 
https://packagist.org/search/?tags=robots.txt

BrocksiNet replied:
Sure would work. Will think about it.

BrocksiNet replied:
What do you think about:
https://github.com/bopoda/robots-txt-parser/tree/master/src/RobotsTxtParser

BrocksiNet replied:
I think a library is too much here. Also, the libraries I checked are not really maintained. We only have 86 Lines of Code.

mitelg commented on src/Storefront/Page/Robots/Parser/RobotsDirectiveParser.php:
```diff
@@ -0,0 +1,102 @@
+
+            // Validate directive
+            if (!\in_array($normalizedType, self::KNOWN_DIRECTIVES, true)) {
+                continue;
```
instead of ignoring, should we maybe log a warning or something?

mitelg commented on src/Storefront/Page/Robots/Parser/RobotsDirectiveParser.php:
```diff
@@ -0,0 +1,102 @@
+            $directiveType = ucfirst($normalizedType);
+
+            // Handle User-agent directive
+            if ($normalizedType === 'user-agent') {
```
please add constants or an enum for the types

mitelg commented on src/Storefront/Page/Robots/RobotsPageLoader.php:
```diff
@@ -69,39 +77,109 @@ private function getDomains(string $hostname, Context $context): SalesChannelDom
-        return $domainRuleCollection;
+        // Build final global blocks with merged path directives
+        foreach ($globalBlocksByHash as $data) {
+            /** @var RobotsUserAgentBlock $block */
```
```suggestion
```

no var annotations

mitelg commented on src/Storefront/Page/Robots/Struct/DomainRuleStruct.php:
```diff
@@ -13,9 +14,22 @@ class DomainRuleStruct extends Struct
      */
     private array $rules = [];
```
should we deprecate this property as well?

mitelg commented on src/Storefront/Page/Robots/Struct/RobotsDirective.php:
```diff
@@ -0,0 +1,39 @@
+use Shopware\Core\Framework\Struct\Struct;
+
+#[Package('framework')]
+class RobotsDirective extends Struct
```
why extending from `Struct`?

mitelg commented on src/Storefront/Page/Robots/Struct/RobotsDirective.php:
```diff
@@ -0,0 +1,39 @@
+
+    public static function isPathBased(string $type): bool
+    {
+        return \in_array(mb_strtolower($type), ['allow', 'disallow'], true);
```
same here, please introduce constants or an enum for all types

mitelg commented on src/Storefront/Page/Robots/Struct/RobotsUserAgentBlock.php:
```diff
@@ -0,0 +1,65 @@
+use Shopware\Core\Framework\Util\Hasher;
+
+#[Package('framework')]
+class RobotsUserAgentBlock extends Struct
```
why extending `Struct`?

mitelg commented on src/Storefront/Page/Robots/Struct/RobotsDirective.php:
```diff
@@ -0,0 +1,39 @@
+    ) {
+    }
+
+    public static function isPathBased(string $type): bool
```
why is this static? if it would be not, you could simply call the method without pass the type :thinking:

mitelg commented on src/Storefront/Page/Robots/Struct/RobotsUserAgentBlock.php:
```diff
@@ -0,0 +1,65 @@
+    {
+        $parts = [$this->userAgent];
+        foreach ($this->getNonPathDirectives() as $directive) {
+            $parts[] = $directive->type . ':' . $directive->value;
```
is it necessary to put the parts together like this? couldn't you simply use the raw values you have in the properties?

mitelg commented on tests/unit/Storefront/Page/Robots/RobotsPageLoaderTest.php:
```diff
@@ -198,4 +150,547 @@ public function testLoadWithMultipleDomains(): void
+    /**
+     * Common assertions for basic page structure
+     */
+    private function assertBasicPageStructure(\Shopware\Storefront\Page\Robots\RobotsPage $page, int $expectedSitemaps, int $expectedDomainRules, int $expectedGlobalBlocks): void
```
please import the usages and add some line breaks, if it gets to long

BrocksiNet committed:
Merge branch 'trunk' into feat/improve-robots-txt

BrocksiNet committed:
fix: review improvements

BrocksiNet committed:
Merge branch 'trunk' into feat/improve-robots-txt

mitelg commented on src/Storefront/Page/Robots/Struct/DomainRuleStruct.php:
```diff
@@ -21,6 +21,8 @@ class DomainRuleStruct extends Struct
     /**
      * @param ParsedRobots|string $rules The robots.txt rules as string (legacy) or parsed object (new)
+     *
+     * @deprecated 6.7.4.0 - Passing a string as the first parameter is deprecated. Pass a ParsedRobots object instead.
```
please review your AI stuff, really!

BrocksiNet replied:
This was already removed

mitelg commented on src/Storefront/Page/Robots/Struct/RobotsDirective.php:
```diff
@@ -3,25 +3,29 @@
+class RobotsDirective
 {
     public function __construct(
         public readonly string $type,
```
why not expecting the enum here?

BrocksiNet committed:
fix: more review improvements

mitelg commented on src/Storefront/Page/Robots/Struct/RobotsDirectiveType.php:
```diff
@@ -0,0 +1,55 @@
+     *
+     * @return list<self>
+     */
+    public static function getPathBased(): array
```
seems to be unused

mitelg commented on src/Storefront/Page/Robots/Struct/RobotsDirectiveType.php:
```diff
@@ -0,0 +1,55 @@
+     */
+    public static function tryFromInsensitive(string $value): ?self
+    {
+        $normalized = ucfirst(mb_strtolower($value));
```
```suggestion
        return self::tryFrom(ucfirst(mb_strtolower($value)));
```

`self::tryFrom` is pseudo code, not tested, but it should not be necessary to iteratate over the cases and use the built-in function

mitelg commented on src/Storefront/Page/Robots/Struct/RobotsUserAgentBlock.php:
```diff
@@ -52,13 +51,18 @@ public function render(): string
+        $parts = [
+            $this->userAgent,
+            ...array_map(
+                static fn (RobotsDirective $d) => $d->type . ':' . $d->value,
```
my question was more like, why is it necessary to put a string together here?

BrocksiNet committed:
fix: bring back the ignore of optional for bc

BrocksiNet committed:
fix: add enum to RobotsDirective

BrocksiNet committed:
fix: template output

BrocksiNet committed:
fix: improve RobotsDirectiveType

BrocksiNet committed:
Merge branch 'trunk' into feat/improve-robots-txt

BrocksiNet committed:
refactor: Update robots.txt handling and deprecate old methods

- Enhanced help text for the robots.txt configuration to clarify the impact of disabling default rules.
- Refactored `RobotsPageLoader` and `ParsedRobots` to access user agent blocks and orphaned path directives directly, removing deprecated getter methods.
- Marked `getRules` and `getUserAgentBlocks` methods in `DomainRuleStruct` as deprecated, suggesting the use of `getDirectives` instead.
- Updated unit tests to reflect changes in method access and ensure functionality remains intact.

BrocksiNet committed:
feat: Enhance robots.txt parsing with error handling and extensibility

- Introduced `RobotsDirectiveParsingEvent` and `RobotsUnknownDirectiveEvent` to allow developers to modify parsed results and handle custom directives.
- Added error tracking for malformed lines and unknown directives in robots.txt, with logging of issues upon saving configuration.
- Updated `RobotsDirectiveParser` to support extensibility events and improved parsing logic.
- Introduced new value objects for parsed robots.txt data, including `ParseIssue` and `ParseIssueSeverity`.
- Added a new subscriber, `RobotsConfigChangeSubscriber`, to log parsing issues when robots.txt is saved.
- Updated unit tests to cover new functionality and ensure robust error handling.

BrocksiNet committed:
docs: Add coverage ignore comment for ParseIssueSeverity enum

- Added a comment to the `ParseIssueSeverity` enum to indicate that it is a simple enum with no logic, and is covered by tests for `ParseIssue` and `ParsedRobots`.

BrocksiNet committed:
Merge branch 'trunk' into feat/improve-robots-txt

BrocksiNet committed:
feat: Deprecate string format in DomainRuleStruct constructor

- Deprecated passing a string to the `DomainRuleStruct` constructor; users should now utilize the `ParsedRobots` object for improved parsing and validation.
- Updated documentation to reflect the deprecation and provided migration guidance for future versions.
- Enhanced unit tests to cover the deprecated string format, marking them for removal in v6.8.0.

mitelg commented on changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md:
```diff
@@ -0,0 +1,243 @@
+* Added `Shopware\Storefront\Page\Robots\Parser\RobotsDirectiveParser` to parse robots.txt with error tracking and extensibility events
+* Added `Shopware\Storefront\Page\Robots\Event\RobotsDirectiveParsingEvent` to allow modification of parsed results
+* Added `Shopware\Storefront\Page\Robots\Event\RobotsUnknownDirectiveEvent` to handle custom directives
+* Added robots.txt parsing value objects: `ParsedRobots`, `ParseIssue`, `ParseIssueSeverity`, `RobotsDirective`, `RobotsDirectiveType`, `RobotsUserAgentBlock`
```
probably not relevant to mention. those are internals which are probably not used directly by externals

mitelg commented on changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md:
```diff
@@ -0,0 +1,243 @@
+* Added `Shopware\Storefront\Page\Robots\Event\RobotsDirectiveParsingEvent` to allow modification of parsed results
+* Added `Shopware\Storefront\Page\Robots\Event\RobotsUnknownDirectiveEvent` to handle custom directives
+* Added robots.txt parsing value objects: `ParsedRobots`, `ParseIssue`, `ParseIssueSeverity`, `RobotsDirective`, `RobotsDirectiveType`, `RobotsUserAgentBlock`
+* Added `Shopware\Storefront\Page\Robots\RobotsConfigChangeSubscriber` to log parsing issues when robots.txt is saved
```
probably also not really relevant for externals

mitelg commented on changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md:
```diff
@@ -0,0 +1,243 @@
+* Added `Shopware\Storefront\Page\Robots\Event\RobotsUnknownDirectiveEvent` to handle custom directives
+* Added robots.txt parsing value objects: `ParsedRobots`, `ParseIssue`, `ParseIssueSeverity`, `RobotsDirective`, `RobotsDirectiveType`, `RobotsUserAgentBlock`
+* Added `Shopware\Storefront\Page\Robots\RobotsConfigChangeSubscriber` to log parsing issues when robots.txt is saved
+* Added system configuration option `core.basicInformation.robotsDisableDefaults` to disable default robots.txt rules
```
instead of mentioning a new config with the technical name I would add something to the upgrade information, mentioning a new option in the admin

mitelg replied:
I just saw, that this is already mentioned, so this can be removed.

mitelg commented on changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md:
```diff
@@ -0,0 +1,243 @@
+
+___
+# Storefront
+* Changed rendering of robots.txt to support custom User-agent blocks
```
```suggestion
* Changed rendering of `robots.txt` file to support custom User-agent blocks
```

mitelg commented on changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md:
```diff
@@ -0,0 +1,243 @@
+___
+# Storefront
+* Changed rendering of robots.txt to support custom User-agent blocks
+* Changed robots.txt template to respect new `robotsDisableDefaults` configuration
```
imo, also not really relevant

mitelg commented on changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md:
```diff
@@ -0,0 +1,243 @@
+author_github: BrocksiNet
+---
+# Core
+* Added `Shopware\Storefront\Page\Robots\Parser\RobotsDirectiveParser` to parse robots.txt with error tracking and extensibility events
```
```suggestion
* Added `Shopware\Storefront\Page\Robots\Parser\RobotsDirectiveParser` to parse `robots.txt` file with error tracking and extensibility events
```

mitelg commented on changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md:
````diff
@@ -0,0 +1,243 @@
+new DomainRuleStruct($parsedRobots, '/en');
+```
+
+**Migration required**: The string format is deprecated and will be removed in v6.8.0. Use `RobotsDirectiveParser::parse()` instead to get validation, error logging, and event support.
````
```suggestion
Passing a string to the `DomainRuleStruct` is deprecated and will be removed with the next major version.
Use `RobotsDirectiveParser::parse()` instead to get validation, error logging, and event support.
```

mitelg commented on changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md:
```diff
@@ -0,0 +1,243 @@
+User-agent: *
+Allow: /
+Disallow: /*?
+Allow: /*theme/
```
where does this media stuff come from? seems unrelated and is a bit confusing

mitelg commented on changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md:
```diff
@@ -0,0 +1,243 @@
+**Logging behavior:**
+- Issues are logged only when configuration is saved via admin (not on every page load)
+- Each issue includes line number, line content, and error reason
+- Errors are logged with `error` level, warnings with `warning` level
```
![](https://i.kym-cdn.com/entries/icons/original/000/000/015/oreally.jpg)

mitelg commented on changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md:
```diff
@@ -0,0 +1,243 @@
+
+        if (!$hasAdminDisallow) {
+            // Add a warning about missing security directive
+            $newIssues = [
```
that is not really good dev ex :thinking: 

should be more something like this: 
```
$event->parsedResult->issues->add(new ParseIssue(0, '', 'Security: /admin/ should be disallowed', ParseIssueSeverity::WARNING))
```

mitelg commented on changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md:
```diff
@@ -0,0 +1,243 @@
+___
+# Upgrade Information
+
+## robots.txt Configuration Enhancement
```
the documentation of the new feature is very extensive. which is good, don't get me wrong. 
but maybe it would be better to prepare a documentation article first and just link it here in the changelog. wdyt?

Phil23 replied:
Documentation should be done. But in terms of the new changelog process this should also go into the release notes, right?

mitelg commented on src/Core/System/Resources/config/basicInformation.xml:
```diff
@@ -204,12 +204,19 @@
+            <name>robotsDisableDefaults</name>
+            <label>Disable default robots.txt rules</label>
+            <label lang="de-DE">Standard-Regeln für robots.txt deaktivieren</label>
+            <helpText>When enabled, Shopware's standard built-in robots.txt rules such as "User-agent: *", "Allow: /", "Disallow: /*?", etc. will not be included in your robots.txt. However, your own custom rules from the configuration below will still be included.</helpText>
```
```suggestion
            <helpText>If set, Shopware's standard built-in robots.txt rules such as "User-agent: *", "Allow: /", "Disallow: /*?", etc. will not be included in your robots.txt. Instead only your own custom rules from the configuration below are considered.</helpText>
```

maybe a bit clearer?

mitelg replied:
German text needs to be adjusted accordingly, if you agree

mitelg commented on src/Storefront/Page/Robots/Event/RobotsDirectiveParsingEvent.php:
```diff
@@ -0,0 +1,71 @@
+ *
+ *         // Add custom validation logic
+ *         if ($this->hasInvalidConfiguration($result)) {
+ *             $event->addIssue(new ParseIssue(...));
```
example is not in sync with the class and the changelog :grimacing:

mitelg replied:
one reason more, why this should only be documented in one place

mitelg commented on src/Storefront/Page/Robots/Event/RobotsDirectiveParsingEvent.php:
```diff
@@ -0,0 +1,71 @@
+class RobotsDirectiveParsingEvent extends Event implements ShopwareEvent
+{
+    public function __construct(
+        private readonly string $text,
```
make them public properties without setter and getter

mitelg commented on src/Storefront/Page/Robots/Event/RobotsDirectiveParsingEvent.php:
````diff
@@ -0,0 +1,71 @@
+ * ```
+ */
+#[Package('framework')]
+class RobotsDirectiveParsingEvent extends Event implements ShopwareEvent
````
are you sure it should be extending the Symfony Event? :thinking: I mean is it intended that listeners can stop propagation?

BrocksiNet replied:
Most of our events can be stopped. I just follow our regular pattern here. Theoretically, it can be stopped, but practically, most of the time, nobody will stop it.

mitelg replied:
the question is: do we want to give people the possibility to stop it?

mitelg commented on src/Storefront/Page/Robots/Event/RobotsUnknownDirectiveEvent.php:
```diff
@@ -0,0 +1,113 @@
+    private ?ParseIssue $issue = null;
+
+    public function __construct(
+        private readonly int $lineNumber,
```
make them public without getters

mitelg commented on src/Storefront/Page/Robots/Parser/RobotsDirectiveParser.php:
```diff
@@ -0,0 +1,146 @@
+                // We're in a user-agent block
+                $currentDirectives[] = $directive;
+            } else {
+                // Orphaned directive (backward compatibility)
```
is that something, we want to deprecate then?

mitelg commented on src/Storefront/Page/Robots/RobotsConfigChangeSubscriber.php:
```diff
@@ -0,0 +1,76 @@
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            SystemConfigChangedEvent::class => 'onSystemConfigChanged',
```
you probably choose this event instead of `RobotsDirectiveParsingEvent` so it is not done on every request?

mitelg commented on src/Storefront/Page/Robots/Parser/RobotsDirectiveParser.php:
```diff
@@ -0,0 +1,146 @@
+        $issues = [];
+        $lineNumber = 0;
+
+        foreach ($lines as $line) {
```
the foreach is very long, would be nice, if it would be possible to create some private method calls

mitelg commented on src/Storefront/Page/Robots/RobotsPage.php:
```diff
@@ -16,6 +20,11 @@ class RobotsPage extends Struct
     protected array $sitemaps;
 
+    /**
+     * @var RobotsUserAgentBlock[]
```
```suggestion
     * @var list<RobotsUserAgentBlock>
```

mitelg commented on src/Storefront/Page/Robots/RobotsPage.php:
```diff
@@ -41,4 +50,20 @@ public function setSitemaps(array $sitemaps): void
     }
+
+    /**
+     * @return RobotsUserAgentBlock[]
```
```suggestion
     * @return list<RobotsUserAgentBlock>
```

mitelg commented on src/Storefront/Page/Robots/RobotsPage.php:
```diff
@@ -41,4 +50,20 @@ public function setSitemaps(array $sitemaps): void
+    }
+
+    /**
+     * @param RobotsUserAgentBlock[] $blocks
```
```suggestion
     * @param list<RobotsUserAgentBlock> $blocks
```

mitelg commented on src/Storefront/Page/Robots/RobotsPage.php:
```diff
@@ -16,6 +20,11 @@ class RobotsPage extends Struct
+    /**
+     * @var RobotsUserAgentBlock[]
+     */
+    protected array $globalUserAgentBlocks = [];
```
I know, we have a pattern within those Pages, but maybe we can also start using them without getter and setter?

mitelg commented on src/Storefront/Page/Robots/RobotsPageLoader.php:
```diff
@@ -69,39 +77,108 @@ private function getDomains(string $hostname, Context $context): SalesChannelDom
+     * @return array<string, SalesChannelDomainEntity> Array keyed by domain hostname with selected domain entities
      */
-    private function getDomainRules(string $hostname, SalesChannelDomainCollection $domains): DomainRuleCollection
+    private function selectDomainsByHostname(SalesChannelDomainCollection $domains, string $hostname): array
```
please sort the private methods, after when they are called

mitelg replied:
why not returning a filtered `SalesChannelDomainCollection`?

mitelg commented on src/Storefront/Page/Robots/RobotsPageLoader.php:
```diff
@@ -69,39 +77,108 @@ private function getDomains(string $hostname, Context $context): SalesChannelDom
     {
-        $domainRuleCollection = new DomainRuleCollection();
+        $selectedDomains = [];
+        \assert($hostname !== '');
```
why do you assert that, if you already declared it as non-empty on the PHPDoc?

BrocksiNet replied:
that's a little more security, not a big deal

mitelg commented on src/Storefront/Page/Robots/RobotsPageLoader.php:
```diff
@@ -69,39 +77,108 @@ private function getDomains(string $hostname, Context $context): SalesChannelDom
+     *
+     * @param non-empty-string $hostname
+     *
+     * @return array{0: RobotsUserAgentBlock[], 1: DomainRuleCollection}
```
```suggestion
     * @return array{0: list<RobotsUserAgentBlock>, 1: DomainRuleCollection}
```

mitelg commented on src/Storefront/Page/Robots/RobotsPageLoader.php:
```diff
@@ -40,9 +44,13 @@ public function load(Request $request, Context $context): RobotsPage
 
-            $page->setDomainRules($this->getDomainRules($hostname, $domains));
-            $page->setSitemaps($this->getSitemaps($domains));
+            [$globalBlocks, $domainRules] = $this->collectRules($hostname, $domains, $context);
```
I don't like methods returning multiple things, but in this case it makes sense, right? as we need to iterate the same things for both?

BrocksiNet replied:
yes we need that

mitelg commented on src/Storefront/Page/Robots/Struct/DomainRuleStruct.php:
```diff
@@ -2,49 +2,109 @@
 
-    public function __construct(string $rules, private readonly string $basePath)
+    /**
+     * @var RobotsDirective[]
```
```suggestion
     * @var list<RobotsDirective>
```

mitelg commented on src/Storefront/Page/Robots/Struct/DomainRuleStruct.php:
```diff
@@ -2,49 +2,109 @@
     }
 
+    /**
+     * @return RobotsDirective[]
```
```suggestion
     * @return list<RobotsDirective>
```

mitelg commented on src/Storefront/Page/Robots/Struct/DomainRuleStruct.php:
```diff
@@ -2,49 +2,109 @@
+        foreach ($parsed->orphanedPathDirectives as $directive) {
+            $directiveWithPath = $directive->withBasePath($this->basePath);
+            $this->directives[] = $directiveWithPath;
+            $this->rules[] = ['type' => $directiveWithPath->type->value, 'path' => $directiveWithPath->value];
```
please surround this with a if condition and feature check

mitelg commented on src/Storefront/Page/Robots/Struct/DomainRuleStruct.php:
```diff
@@ -2,49 +2,109 @@
+            foreach ($block->getPathDirectives() as $directive) {
+                $directiveWithPath = $directive->withBasePath($this->basePath);
+                $this->directives[] = $directiveWithPath;
+                $this->rules[] = ['type' => $directiveWithPath->type->value, 'path' => $directiveWithPath->value];
```
please surround this with a if condition and feature check

BrocksiNet replied:
not needed because we introduced `ParsedRobots` object, see the if/else in __construct, should be clear how to clean up later

mitelg commented on src/Storefront/Page/Robots/Struct/DomainRuleStruct.php:
```diff
@@ -2,49 +2,109 @@
-            $this->rules[] = ['type' => ucfirst($ruleType), 'path' => '/' . ltrim($path, '/')];
+            $normalizedPath = '/' . ltrim($path, '/');
+
+            $this->rules[] = ['type' => $directiveType->value, 'path' => $normalizedPath];
```
please surround this with a if condition and feature check

mitelg commented on src/Storefront/Page/Robots/Struct/DomainRuleStruct.php:
```diff
@@ -2,49 +2,109 @@
+        }
+    }
+
+    private function parseRulesFromString(string $rules): void
```
calling the parser here is too much? :thinking: maybe we could make some logic public in the parser, so it could be re-used here? :thinking: 

and if we make something public, deprecating it directly to be private again in 6.8 :grin:

mitelg commented on src/Storefront/Page/Robots/Struct/RobotsUserAgentBlock.php:
```diff
@@ -0,0 +1,67 @@
+class RobotsUserAgentBlock
+{
+    /**
+     * @param RobotsDirective[] $directives
```
```suggestion
     * @param list<RobotsDirective> $directives
```

mitelg commented on src/Storefront/Page/Robots/Struct/RobotsUserAgentBlock.php:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    /**
+     * @return RobotsDirective[]
```
```suggestion
     * @return list<RobotsDirective>
```

mitelg commented on src/Storefront/Page/Robots/Struct/RobotsUserAgentBlock.php:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    /**
+     * @return RobotsDirective[]
```
```suggestion
     * @return list<RobotsDirective>
```

mitelg commented on src/Storefront/Resources/views/storefront/page/robots/robots.txt.twig:
```diff
@@ -1,37 +1,54 @@
     {% block robots_txt_content %}
         {% block robots_txt_content_user_agent %}
-            User-agent: *
+            {% if not config('core.basicInformation.robotsDisableDefaults') %}
```
we should put that into a variable so the config service is not called multiple times here

mitelg commented on src/Storefront/Resources/views/storefront/page/robots/robots.txt.twig:
```diff
@@ -1,37 +1,54 @@
 
+            {% if page.globalUserAgentBlocks|length > 0 %}
+                {% block robots_txt_content_global_user_agent_blocks %}
+                    {% for block in page.globalUserAgentBlocks %}
```
is `block` working as variable name? here on GitHub it is highlighted as syntax :grin:

mitelg commented on tests/unit/Storefront/Page/Robots/Event/RobotsUnknownDirectiveEventTest.php:
```diff
@@ -0,0 +1,143 @@
+#[CoversClass(RobotsUnknownDirectiveEvent::class)]
+class RobotsUnknownDirectiveEventTest extends TestCase
+{
+    public function testGettersReturnConstructorValues(): void
```
senseless test. make the properties public and remove that test :grimacing:

mitelg commented on tests/unit/Storefront/Page/Robots/Parser/ParseIssueTest.php:
```diff
@@ -0,0 +1,57 @@
+ * @internal
+ */
+#[CoversClass(ParseIssue::class)]
+class ParseIssueTest extends TestCase
```
all tests are senseless...

mitelg commented on tests/unit/Storefront/Page/Robots/Parser/ParsedRobotsTest.php:
```diff
@@ -0,0 +1,264 @@
+#[CoversClass(ParsedRobots::class)]
+class ParsedRobotsTest extends TestCase
+{
+    public function testConstructorAndGetters(): void
```
please no getter and setter tests...

mitelg commented on tests/unit/Storefront/Page/Robots/Parser/ParsedRobotsTest.php:
```diff
@@ -0,0 +1,264 @@
+        static::assertTrue($parsed->hasUserAgentBlocks());
+    }
+
+    public function testWithEmptyArrays(): void
```
the important thing is already tested above in `testHasUserAgentBlocksReturnsFalseWhenNoBlocks`

mitelg commented on tests/unit/Storefront/Page/Robots/Parser/ParsedRobotsTest.php:
```diff
@@ -0,0 +1,264 @@
+
+        $parsed = new ParsedRobots($userAgentBlocks, $orphanedDirectives);
+
+        // Verify the arrays are the same objects (not cloned)
```
why???

mitelg commented on tests/unit/Storefront/Page/Robots/Parser/ParsedRobotsTest.php:
```diff
@@ -0,0 +1,264 @@
+        static::assertSame($orphanedDirectives, $parsed->orphanedPathDirectives);
+    }
+
+    public function testGetUserAgentBlocksReturnsEmptyArrayWhenNoBlocks(): void
```
also already tested

mitelg commented on tests/unit/Storefront/Page/Robots/Parser/ParsedRobotsTest.php:
```diff
@@ -0,0 +1,264 @@
+        static::assertEmpty($blocks);
+    }
+
+    public function testGetOrphanedPathDirectivesReturnsEmptyArrayWhenNoDirectives(): void
```
you can almost delete the whole test stuff. please do not do stupid senseless test. please check what is really done, after your AI generated this sh*t

mitelg commented on tests/unit/Storefront/Page/Robots/Parser/ParsedRobotsTest.php:
```diff
@@ -0,0 +1,264 @@
+        static::assertSame(RobotsDirectiveType::ALLOW, $directives[1]->type);
+    }
+
+    public function testMixedUserAgentBlocksAndOrphanedDirectives(): void
```
also duplicated

mitelg commented on tests/unit/Storefront/Page/Robots/Parser/ParsedRobotsTest.php:
```diff
@@ -0,0 +1,264 @@
+ * @internal
+ */
+#[CoversClass(ParsedRobots::class)]
+class ParsedRobotsTest extends TestCase
```
please remove the whole tests and then check again, that you only test real logic from that class. there is no value in testing getters and setter in 1000 different ways...

BrocksiNet replied:
done ✅

mitelg commented on tests/unit/Storefront/Page/Robots/Parser/RobotsDirectiveParserTest.php:
```diff
@@ -0,0 +1,533 @@
+
+        static::assertCount(2, $result->userAgentBlocks);
+
+        // Both user agents should have the same directive
```
you only check, that you have directives but not, that they are actually the same

BrocksiNet replied:
done ✅

mitelg commented on tests/unit/Storefront/Page/Robots/Parser/RobotsDirectiveParserTest.php:
```diff
@@ -0,0 +1,533 @@
+
+        $result = $this->parser->parse($text, Context::createDefaultContext());
+
+        static::assertCount(1, $result->userAgentBlocks);
```
additionally you could check, that the types are normalised

BrocksiNet replied:
done ✅

mitelg commented on tests/unit/Storefront/Page/Robots/Parser/RobotsDirectiveParserTest.php:
```diff
@@ -0,0 +1,533 @@
+        static::assertSame(RobotsDirectiveType::DISALLOW, $result->orphanedPathDirectives[0]->type);
+    }
+
+    public function testParseMalformedLineWithoutColon(): void
```
same as `testParseIgnoresMalformedLines`

mitelg commented on tests/unit/Storefront/Page/Robots/Parser/RobotsDirectiveParserTest.php:
```diff
@@ -0,0 +1,533 @@
+        static::assertSame(ParseIssueSeverity::ERROR, $error->severity);
+    }
+
+    public function testParseUnknownDirectiveType(): void
```
same as `testParseIgnoresUnknownDirectives`

mitelg reviewed (changes requested):
All in all I think this is a great improvement and the architecture in general is fine :+1: good job Björn and AI :grin: 
It might be a lot of comments, but no real major thing.

I stopped at the tests, because they are partially total bullshit. many stuff is tested multiple times, unnecessary getter and setter testing. 
please revisit those again and create meaningful tests.

BrocksiNet committed:
Merge branch 'trunk' into feat/improve-robots-txt

BrocksiNet committed:
feat(robots.txt): Improve tests and changelog file after review

BrocksiNet committed:
refactor(tests): Simplify ParsedRobotsTest by removing redundant tests and enhancing issue filtering logic

BrocksiNet committed:
feat(robots.txt): Simplify Events, improve Tests, and Changelog

BrocksiNet committed:
Merge branch 'trunk' into feat/improve-robots-txt

BrocksiNet committed:
fix(tests): Update method call to use getSalesChannelId for improved clarity in RobotsPageLoaderTest

BrocksiNet committed:
refactor(tests): Enhance RobotsPageLoaderTest by consolidating tests and adding helper methods for clarity

BrocksiNet committed:
Merge branch 'trunk' into feat/improve-robots-txt

BrocksiNet commented:
PR for updating the docs is here: https://github.com/shopware/docs/pull/1988

After release, we should also update this page:
https://docs.shopware.com/en/shopware-6-en/tutorials-and-faq/creation-of-robots-txt (managed in SBP)

Phil23 commented on src/Core/System/Resources/config/basicInformation.xml:
```diff
@@ -204,12 +204,19 @@
+            <label>Domain specific robots.txt rules</label>
+            <label lang="de-DE">Domain spezifische robots.txt Regeln</label>
+            <helpText>Define robots.txt rules. Supports User-agent blocks (applied globally across all domains) and Allow/Disallow rules (applied to this domain path). Example: "User-agent: Googlebot&#10;Crawl-delay: 10&#10;Disallow: /private/"</helpText>
+            <helpText lang="de-DE">Definieren Sie robots.txt-Regeln. Unterstützt User-agent-Blöcke (global für alle Domains) und Allow/Disallow-Regeln (für diesen Domain-Pfad). Beispiel: "User-agent: Googlebot&#10;Crawl-delay: 10&#10;Disallow: /private/"</helpText>
```
We use "Du" for salutations in German across the administration, if I am not mistaken.

BrocksiNet committed:
Merge branch 'trunk' into feat/improve-robots-txt

BrocksiNet committed:
feat: Update release and upgrade files, remove changelog

BrocksiNet committed:
fix: Update help text for cookie consent and robots.txt configuration in basicInformation.xml

mitelg commented on RELEASE_INFO-6.7.md:
```diff
@@ -5,15 +5,18 @@
 
 ## Core
+* Added `Shopware\Storefront\Page\Robots\Parser\RobotsDirectiveParser` to parse `robots.txt` files with error tracking and extensibility events.
+* The constructor of `Shopware\Storefront\Page\Robots\Struct\DomainRuleStruct` is now deprecated for string parameters. Use a `ParsedRobots` object instead. This will be removed in v6.8.0.
```
```suggestion
* The constructor of `Shopware\Storefront\Page\Robots\Struct\DomainRuleStruct` is now deprecated for string parameters and support for this will be removed with the next major version. Use a `ParsedRobots` object instead.
```

mitelg commented on UPGRADE-6.8.md:
```diff
@@ -1,4 +1,19 @@
+
+## Removal of string parameter in `DomainRuleStruct` constructor
+
+The deprecated string parameter in the `Shopware\Storefront\Page\Robots\Struct\DomainRuleStruct` constructor will be removed. Use `Shopware\Storefront\Page\Robots\Parser\RobotsDirectiveParser::parse()` to create a `ParsedRobots` object instead.
```
```suggestion
The deprecated string parameter in the `Shopware\Storefront\Page\Robots\Struct\DomainRuleStruct` constructor was removed. 
Use `Shopware\Storefront\Page\Robots\Parser\RobotsDirectiveParser::parse()` to create a `ParsedRobots` object instead.
```

BrocksiNet committed:
Merge branch 'trunk' into feat/improve-robots-txt

BrocksiNet committed:
fix: only use ShopwareEvent

BrocksiNet committed:
test: add user agent blocks validation in ParsedRobotsTest

BrocksiNet committed:
chore: update release and upgrade notes for DomainRuleStruct deprecation

mitelg commented on RELEASE_INFO-6.7.md:
```diff
@@ -6,15 +6,18 @@
 ## API
 
 ## Core
+* Added `Shopware\Storefront\Page\Robots\Parser\RobotsDirectiveParser` to parse `robots.txt` files with error tracking and extensibility events.
```
please add a headline and remove the bullet points and write a text instead

mitelg commented on RELEASE_INFO-6.7.md:
```diff
@@ -6,15 +6,18 @@
 ## Administration
 
 ## Storefront
+* The rendering of the `robots.txt` file has been changed to support custom `User-agent` blocks and the full `robots.txt` standard. For a detailed guide on how to use the new features and extend the functionality, please refer to our documentation guide [Extend robots.txt configuration](https://developer.shopware.com/docs/guides/plugins/plugins/content/seo/extend-robots-txt.html).
```
please add a headline and remove the bullet points and write a text instead

mitelg replied:
should maybe also added to the `Features` section?

mitelg commented on UPGRADE-6.8.md:
```diff
@@ -114,6 +130,7 @@ The constructor of the `EntityDefinition` has been removed, therefore the call o
  {
      // snip
 
+
```
merge foo here and all changes below

mitelg commented on UPGRADE-6.8.md:
```diff
@@ -1,4 +1,20 @@
 # 6.8.0.0
+
+## Removal of string parameter in `DomainRuleStruct` constructor
```
we should wait until https://github.com/shopware/shopware/pull/13346 is merged with the new structure

BrocksiNet committed:
Merge branch 'trunk' into feat/improve-robots-txt

BrocksiNet committed:
chore: update RELEASE_INFO and UPGRADE notes for robots.txt enhancements and DomainRuleStruct changes

BrocksiNet committed:
chore: enhance RELEASE_INFO with details on new RobotsDirectiveParser and DomainRuleStruct deprecation

BrocksiNet committed:
chore: update RELEASE_INFO with new robots.txt configuration details and enhancements

BrocksiNet committed:
chore: update UPGRADE-6.8.md with removal of DomainRuleStruct string parameter and related changes

BrocksiNet committed:
refactor: update type hints for user agent blocks and restore selectDomainsByHostname method

- Changed type hints for globalUserAgentBlocks and related methods to use list<RobotsUserAgentBlock> for better type safety.
- Restored the selectDomainsByHostname method to maintain functionality after its removal, ensuring HTTPS preference is handled correctly.
- Minor comment adjustment in DomainRuleStruct for clarity.

BrocksiNet committed:
refactor: update type hints in DomainRuleStruct and RobotsUserAgentBlock for improved type safety

- Changed type hints from RobotsDirective[] to list<RobotsDirective> in DomainRuleStruct and RobotsUserAgentBlock.
- Updated return types in relevant methods to reflect the new list type for better clarity and consistency.

BrocksiNet committed:
chore: add code coverage ignore annotations and remove obsolete tests

- Added `@codeCoverageIgnore` annotations to `RobotsUnknownDirectiveEvent` and `ParseIssue` classes to indicate they are simple DTOs with no business logic.
- Removed obsolete unit tests for `RobotsUnknownDirectiveEvent` and `ParseIssue` as they are no longer needed.

BrocksiNet committed:
feat: introduce deprecation for parseDirectiveFromString method and enhance DomainRuleStruct parsing logic

- Added a deprecation notice for the `parseDirectiveFromString` method in `RobotsDirectiveParser`, indicating it will be removed in version 6.8.0.0 due to the deprecation of string parsing.
- Refactored `DomainRuleStruct` to utilize the new `parseDirectiveFromString` method for improved directive parsing and streamlined the collection of directives from user-agent blocks.
- Updated the robots.txt Twig template to use a variable for disabling default rules, enhancing readability and maintainability.

BrocksiNet committed:
Merge branch 'trunk' into feat/improve-robots-txt

BrocksiNet committed:
fix: add missing newline in robots.txt.twig for improved formatting

BrocksiNet committed:
chore: update deprecation notice for parseDirectiveFromString method

- Adjusted the deprecation tag for the `parseDirectiveFromString` method in `RobotsDirectiveParser` to reflect the upcoming removal in version 6.8.0, aligning with the deprecation of string parsing.

BrocksiNet committed:
Merge branch 'trunk' into feat/improve-robots-txt

mitelg commented on src/Storefront/Page/Robots/Parser/RobotsDirectiveParser.php:
```diff
@@ -143,4 +144,26 @@ public function parse(string $text, Context $context, ?string $salesChannelId =
+    {
+        Feature::triggerDeprecationOrThrow(
+            'v6.8.0.0',
+            Feature::deprecatedMethodMessage(self::class, __METHOD__, 'v6.8.0.0', 'will be removed, as string parsing is deprecated')
```
this will create a message like this:

`Use "will be removed, as string parsing is deprecated" instead.`

mitelg commented on src/Storefront/Page/Robots/Parser/RobotsDirectiveParser.php:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    /**
+     * @deprecated tag:v6.8.0 - reason:becomes-obsolete - will be removed, as string parsing is deprecated
```
`reason:becomes-obsolete` does not exists as exception for the deprecation PHPStan rule. 

the `reason:*` stuff is only used for methods, where we cannot add a call of `Feature::triggerDeprecationOrThrow(`, but which we have here. so this can simply be omitted. Teach that your AI please :wink: 

```suggestion
     * @deprecated tag:v6.8.0 - Will be removed with next major, as string parsing is deprecated
```

BrocksiNet replied:
Done. Also, the Rules have been updated.

BrocksiNet committed:
chore: update deprecation message for parseDirectiveFromString method

BrocksiNet committed:
chore: clarify deprecation message for parseDirectiveFromString method
